### PR TITLE
feat(compiler): interface declaration

### LINKED ISSUE #123: Interface Declaration
https://github.com/winglang/wing/blob/main/docs/04-reference/winglang-spec.md#34-interfaces

### PR DESCRIPTION
**Pull request by staycoolcall911**
This PR adds support for `interface` according to the [spec](https://docs.winglang.io/reference/spec#34-interfaces).

Also in this PR:
- [x] Removed interface fields from the spec and our codebase ([slack discussion](https://winglang.slack.com/archives/C048QCN2XLJ/p1680077416200329)).
- [x] Removed access modifiers from interface methods (to comply with the spec).
- [x] (Very) small refactoring in our grammar.js to rename method return type from `type` to `return_type`.
- [x] Removed unused parameter `is_class` from SymbolEnv::new

Fixes #123

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/parser.rs [hunk 1]
```diff
@@ -584,6 +584,93 @@ impl<'s> Parser<'s> {
...
+
+			// ignore commas
+			if !extend.is_named() {
+				continue;
+			}
+
+			let interface_type = self.build_udt_annotation(&extend)?;
+			match interface_type {
+				TypeAnnotation::UserDefined(interface_type) => extends.push(interface_type),
+				_ => {
```
Comment (yoav-steinberg) [inline libs/wingc/src/parser.rs:646]: `build_udt_annotation` alaways returns a udt or an error, so there's no need to handle this, we should just panic.

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 2]
```diff
@@ -400,6 +421,46 @@ exports[`inflight_ref_unknown_op.w > stderr 1`] = `
...
 
+exports[`interface.w > stderr 1`] = `
+"error: Unknown symbol \\"IB\\"
+  --> ../../../examples/tests/invalid/interface.w:2:22
+  |
+2 | interface IA extends IB {
+  |                      ^^ Unknown symbol \\"IB\\"
+
+
+error: Interface \\"IA (at ../../../examples/tests/invalid/interface.w:2:11)\\" extends \\"any\\", which is not an interface
```
Comment (yoav-steinberg) [inline tools/hangar/__snapshots__/invalid.ts.snap:432]: This seems like some bug in our type checker, the error should be that `IB` doesn't exist not that `any` is not an interface.

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 3]
```diff
@@ -400,6 +421,46 @@ exports[`inflight_ref_unknown_op.w > stderr 1`] = `
...
+
+
+error: Unknown symbol \\"IDontExist\\"
+   --> ../../../examples/tests/invalid/interface.w:10:26
+   |
+10 | interface IExist extends IDontExist {
+   |                          ^^^^^^^^^^ Unknown symbol \\"IDontExist\\"
+
+
+error: Interface \\"IExist (at ../../../examples/tests/invalid/interface.w:10:11)\\" extends \\"any\\", which is not an interface
```
Comment (yoav-steinberg) [inline tools/hangar/__snapshots__/invalid.ts.snap:442]: Same issue `IDontExist` doesn't exist, but `any` has nothing to do with it.

File: examples/tests/valid/interface.w
[the remark this answers is not in the document]
Comment (staycoolcall911) [reply examples/tests/valid/interface.w:1]: 1. FYI @yoav-steinberg - @Chriscbr decided to support impl `interface`s only by `resource`s at the moment. Not sure if we have an issue for supporting `class`es too

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 2 above)
Comment (staycoolcall911) [reply tools/hangar/__snapshots__/invalid.ts.snap:432]: We actually show 2 errors:
```sh
error: Unknown symbol "IB"
  --> ../../examples/tests/invalid/interface.w:2:22
  |
2 | interface IA extends IB {
  |                      ^^ Unknown symbol "IB"
error: Interface "IA (at ../../examples/tests/invalid/interface.w:2:11)" extends "any", which is not an interface
```
The first one is the one you expected to have.
Let me see if the second one is something we want to remove.

File: examples/tests/invalid/impl_interface.w [hunk 4]
```diff
@@ -17,3 +17,19 @@ resource C impl cloud.Bucket {
...
+interface I2 extends I1 {
+  inflight method_2(x: str): str;
+}
+
+interface I3 extends I2 {
+  method_3(x: Array<num>): Array<num>;
+}
+
+resource r impl I3 {
+  init() {}
```
Comment (Chriscbr) [inline examples/tests/invalid/impl_interface.w:34]: Can we add a comment saying what the error here is?

Comment (staycoolcall911) [reply examples/tests/invalid/impl_interface.w:37]: Yes - good one. Added

**Review by yoav-steinberg — APPROVED** [review]
I think the invalid snapshot tests still need to be updated (not to include the `any` errors).

File: tools/hangar/__snapshots__/invalid.ts.snap (hunk 2 above)
Comment (staycoolcall911) [reply tools/hangar/__snapshots__/invalid.ts.snap]: Done (removed the 2nd error message)

**Pull request merged**